library/item/item_group/src/main/java/org/quiltmc/qsl/itemgroup/mixin/client/CreativeInventoryScreenMixin.java

```diff
@@ -0,0 +1,175 @@
+
+import net.minecraft.client.gui.screen.ingame.AbstractInventoryScreen;
+import net.minecraft.client.gui.screen.ingame.CreativeInventoryScreen;
+import net.minecraft.client.util.math.MatrixStack;
+import net.minecraft.entity.player.PlayerInventory;
+import net.minecraft.item.ItemGroup;
+import net.minecraft.text.Text;
+
+import org.quiltmc.qsl.itemgroup.impl.CreativeGuiExtensions;
+import org.quiltmc.qsl.itemgroup.impl.QuiltCreativePlayerInventoryScreenWidgets;
+
+@Mixin(CreativeInventoryScreen.class)
+public abstract class CreativeInventoryScreenMixin extends AbstractInventoryScreen<CreativeInventoryScreen.CreativeScreenHandler> implements CreativeGuiExtensions {
+	private CreativeInventoryScreenMixin(CreativeInventoryScreen.CreativeScreenHandler screenHandler, PlayerInventory playerInventory, Text title) {
+		super(screenHandler, playerInventory, title);
+	}
+
+	@Shadow
+	protected abstract void setSelectedTab(ItemGroup itemGroup);
+
+	@Shadow
+	public abstract int getSelectedTab();
+
+	/**
+	 * In order to match the behavior of Vanilla where closing and opening the creative inventory brings you to the previously open item group, we also replicate this behavior with the current page
+	 */
+	@Unique
+	private static int quilt$currentPage = 0;
+
+	@Unique
+	private int getPageOffset(int page) {
+		return switch (page) {
+			case 0 -> 0;
+			case 1 -> 12;
+			default -> 12 + ((12 - QuiltCreativePlayerInventoryScreenWidgets.ALWAYS_SHOWN_GROUPS.size()) * (page - 1));
+		};
+	}
+
+	@Unique
+	private int getOffsetPage(int offset) {
+		if (offset < 12) {
+			return 0;
+		} else {
+			return 1 + ((offset - 12) / (12 - QuiltCreativePlayerInventoryScreenWidgets.ALWAYS_SHOWN_GROUPS.size()));
+		}
+	}
+
+	@Override
+	public void quilt$nextPage() {
+		if (getPageOffset(quilt$currentPage + 1) >= ItemGroup.GROUPS.length) {
+			return;
+		}
+
+		quilt$currentPage++;
+		quilt$updateSelection();
+	}
+
+	@Override
+	public void quilt$previousPage() {
+		if (quilt$currentPage == 0) {
+			return;
+		}
+
+		quilt$currentPage--;
+		quilt$updateSelection();
+	}
+
+	@Override
+	public boolean quilt$isButtonVisible(QuiltCreativePlayerInventoryScreenWidgets.Type type) {
+		return ItemGroup.GROUPS.length > 12;
+	}
+
+	@Override
+	public boolean quilt$isButtonEnabled(QuiltCreativePlayerInventoryScreenWidgets.Type type) {
+		if (type == QuiltCreativePlayerInventoryScreenWidgets.Type.NEXT) {
+			return !(getPageOffset(quilt$currentPage + 1) >= ItemGroup.GROUPS.length);
+		}
+
+		if (type == QuiltCreativePlayerInventoryScreenWidgets.Type.PREVIOUS) {
+			return quilt$currentPage != 0;
+		}
+
+		return false;
+	}
+
+	@Unique
+	private void quilt$updateSelection() {
+		int minPos = getPageOffset(quilt$currentPage);
+		int maxPos = getPageOffset(quilt$currentPage + 1) - 1;
+		int curPos = getSelectedTab();
+
+		if (curPos < minPos || curPos > maxPos) {
+			setSelectedTab(ItemGroup.GROUPS[getPageOffset(quilt$currentPage)]);
+		}
```

Comment:
to fix two bugs i described earlier i modified this if statement to be
```java
		if ((curPos < minPos || curPos > maxPos) &&
				curPos != searchIndex &&
				curPos != ItemGroup.HOTBAR.getIndex() &&
				curPos != ItemGroup.INVENTORY.getIndex()) {
			setSelectedTab(ItemGroup.GROUPS[curPos < minPos ? minPos : (maxPos == getPageOffset(1) - 1) ? 10 : maxPos]);
		}
```
might want to create something like `boolean QuiltCreativePlayerInventoryScreenWidgets.isAlwaysShow(int groupIndex)`